File: src/lib/engine/generator/placement.ts
import type { BaseComponent } from '@/types/schema';

/**
 * Emit a placement wrapper for a component.
 *
 * Absolute mode (default): `#place(dx, dy)[#block(...)[body]]`
 * Flow mode: `#block(width: Wmm, clip: false)[body]` — no fixed position,
 * so Typst flows content naturally and expanding components push siblings down.
 */
export function wrapPlacement(
  base: BaseComponent,
  body: string,
  offsetX: number,
  offsetY: number,
  flowMode?: boolean,
  fillWidth?: boolean
): string {
  const parts: string[] = [];

  if (flowMode) {
    if (base.pageBreakBefore) parts.push('#pagebreak(weak: true)\n');

    if (base.type === 'table') {
      // --- Special Native Flow-Pagination Wrapper for Tables ---
      // - Keeps the table in document flow so rows can paginate naturally.
      // - Ignores y: flow layout order determines vertical position.
      // - No height constraint so Typst splits table rows naturally across page boundaries.
      const x = base.x ?? 0;
      const w = base.width ?? 100;
      const leftPad = fillWidth ? 0 : offsetX + x;
      const widthExpr = fillWidth ? '100%' : `${w}mm`;
      const sizedBlock = `#block(width: ${widthExpr}, clip: false)[${body}]`;
      const inner = leftPad > 0 ? `#pad(left: ${leftPad}mm)[${sizedBlock}]` : sizedBlock;
      const aboveVal = (base as any).marginTop != null ? `${(base as any).marginTop}mm` : '0pt';
      const belowVal =
        (base as any).marginBottom != null ? `${(base as any).marginBottom}mm` : '0pt';
      parts.push(`#block(above: ${aboveVal}, below: ${belowVal}, width: 100%)[${inner}]\n`);
      return parts.join('');
    }

    const x = base.x ?? 0;
    const w = base.width ?? 100;
    const h = base.height ?? 10;
    // When inside a grid cell (fillWidth=true), use 100% so the content
    // fills the cell determined by the grid column width.
    const widthExpr = fillWidth ? '100%' : `${w}mm`;
    // Text, columns, checklist, summary-box, and repeater components use auto-height in flow mode:
    // Typst determines height from content, preventing overflow/overlap.
    // Other components (image, spacer, line, etc.) still need an explicit height.
    const type = base.type as string;
    const autoHeight =
      type === 'text' ||
      type === 'table' ||
      type === 'columns' ||
      type === 'checklist' ||
      type === 'summary-box' ||
      type === 'repeater';

    const sizedBlock = autoHeight
      ? `#block(width: ${widthExpr}, clip: false)[${body}]`
      : `#block(width: ${widthExpr}, height: ${h}mm, clip: false)[${body}]`;
    // Inside a grid cell, skip left-padding (x indent) — the grid handles positioning
    const inner = !fillWidth && x > 0 ? `#pad(left: ${x}mm)[${sizedBlock}]` : sizedBlock;
    // Use component margins for spacing. Text defaults to 2pt below if no margin set,
    // preventing the "cramped" look where text blocks stack flush against each other.
    const aboveVal = (base as any).marginTop != null ? `${(base as any).marginTop}mm` : '0pt';
    const belowVal =
      (base as any).marginBottom != null
        ? `${(base as any).marginBottom}mm`
        : autoHeight
          ? '2pt'
          : '0pt';
    const outerHeight = autoHeight ? '' : `, height: ${h}mm`;
    const outerWidth = fillWidth ? '100%' : '100%';
    parts.push(
      `#block(above: ${aboveVal}, below: ${belowVal}, width: ${outerWidth}${outerHeight})[${inner}]\n`
    );
    return parts.join('');
  }

  // --- Absolute Mode ---
  if (base.pageBreakBefore) parts.push('#pagebreak(weak: true)\n#box()\n');
  const x = base.x ?? 0;
  const y = base.y ?? 0;
  const w = base.width ?? 100;
  const h = base.height ?? 20;
  const absX = offsetX + x;
  const absY = offsetY + y;

  parts.push(
    `#place(top + left, dx: ${absX}mm, dy: ${absY}mm)[#block(width: ${w}mm, height: ${h}mm, clip: false)[${body}]]\n`
  );
  return parts.join('');
}

/**
 * Format a font weight value for Typst.
 * Named strings are quoted ("bold"), numeric values are unquoted integers (700).
 * Typst rejects numeric values inside quotes ("700" is invalid; 700 is valid).
 */
export function formatWeight(weight: string | number | undefined, fallback = 'regular'): string {
  const w = weight ?? fallback;
  if (typeof w === 'number') return String(w);
  // CSS uses "normal" but Typst requires "regular"
  const normalized = w === 'normal' ? 'regular' : w;
  if (/^\d+$/.test(normalized)) {
    return normalized;
  }
  return `"${normalized}"`;
}

/** Format a hex color string for Typst (e.g. "#ff0000" → `rgb("#ff0000")`). */
export function formatColor(color: string | undefined): string {
  const c = (color ?? '').trim();
  if (!c) return 'none';
  return c.startsWith('#') ? `rgb("${c}")` : c;
}

/** Escape content for use inside Typst string literals (`"..."`). */
export function escapeStringLiteral(s: string): string {
  return s.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}

/**
 * Resolves or normalizes a font family name for Typst.
 * Handles legacy names or mismatches between UI and binary-parsed Typst names.
 */
export function formatFontFamily(font: string | undefined): string {
  const f = (font ?? '').trim();
  if (!f || f === 'Sarabun') return 'Sarabun';
  // Legacy map: "LINE Seed Sans TH" -> "LINE Seed Sans"
  if (f === 'LINE Seed Sans TH') return 'LINE Seed Sans';
  return f;
}

File: src-wasm/src/generator/placement.rs
use crate::schema::BaseComponent;

/// Fixed-position wrapper using `#place()`. Components sit at exact mm coordinates.
/// Used for absolute-positioned zones (header, footer, body in non-flow mode).
pub fn wrap_placement(
    base: &BaseComponent,
    body: &str,
    offset_x: &str,
    offset_y: &str,
    prefix: &str,
) -> String {
    let x = base.x.unwrap_or(0.0);
    let y = base.y.unwrap_or(0.0);
    let w = base.width.unwrap_or(100.0);
    let h = base.height.unwrap_or(20.0);

    let off_x: f64 = offset_x.trim_end_matches("mm").trim().parse().unwrap_or(0.0);
    let off_y: f64 = offset_y.trim_end_matches("mm").trim().parse().unwrap_or(0.0);

    let mut out = String::new();
    if base.page_break_before.unwrap_or(false) {
        out.push_str("#pagebreak(weak: true)\n#box()\n");
    }
    out.push_str(&format!(
        "{}place(top + left, dx: {}mm, dy: {}mm)[#block(width: {}mm, height: {}mm, clip: false)[{}]]\n",
        prefix,
        off_x + x,
        off_y + y,
        w,
        h,
        body
    ));
    out
}

/// Flow wrapper — keeps content in document flow (no `#place()`).
/// - `above/below: 0pt` removes Typst's default inter-block spacing so rows stack flush.
/// - Explicit `height` ensures percentage-height children (e.g. `height: 100%` in images)
///   resolve against the component's mm value instead of the full page height.
/// - `y` is intentionally ignored: in flow mode, order is determined by array position,
///   not by absolute coordinates.
/// - `fill_width`: when true, uses `width: 100%` for the inner block (used inside grid cells
///   where the cell width determines the available space).
/// - `is_text`: when true, omits height so Typst auto-sizes based on content (prevents overlap).
pub fn wrap_flow_block(base: &BaseComponent, body: &str, prefix: &str, fill_width: bool, is_text: bool) -> String {
    let x = base.x.unwrap_or(0.0);
    let w = base.width.unwrap_or(190.0);
    let h = base.height.unwrap_or(10.0);

    let mut out = String::new();
    if base.page_break_before.unwrap_or(false) {
        out.push_str("#pagebreak(weak: true)\n");
    }
    if !body.is_empty() {
        // When inside a grid cell (fill_width=true), use 100% so content fills
        // the cell determined by the grid column width.
        let width_expr = if fill_width { "100%".to_string() } else { format!("{}mm", w) };
        // Text: auto height (Typst determines from content) → prevents overlap
        // Other: explicit mm height so percentage-height children resolve correctly
        let inner_block = if is_text {
            format!("#block(width: {}, clip: false)[{}]", width_expr, body)
        } else {
            format!("#block(width: {}, height: {}mm, clip: false)[{}]", width_expr, h, body)
        };
        // Inside a grid cell, skip left-padding (x indent) — the grid handles positioning
        let inner = if !fill_width && x > 0.0 {
            format!("#pad(left: {}mm)[{}]", x, inner_block)
        } else {
            inner_block
        };
        // Use component margins for spacing. Text defaults to 2pt below if no margin set,
        // preventing the "cramped" look where text blocks stack flush against each other.
        let above_val = base.margin_top.map(|m| format!("{}mm", m)).unwrap_or_else(|| "0pt".to_string());
        let below_val = base.margin_bottom.map(|m| format!("{}mm", m)).unwrap_or_else(|| {
            if is_text { "2pt".to_string() } else { "0pt".to_string() }
        });
        // Text: auto outer height too
        let outer_height = if is_text { String::new() } else { format!(", height: {}mm", h) };
        out.push_str(&format!(
            "{}block(above: {}, below: {}, width: 100%{})[{}]\n",
            prefix, above_val, below_val, outer_height, inner
        ));
    }
    out
}

/// Flow wrapper using `#pad` + `#block` for native page-break support.
/// Required for tables so Typst can split rows across pages automatically.
pub fn wrap_flow(
    base: &BaseComponent,
    body: &str,
    offset_x: &str,
    offset_y: &str,
    prefix: &str,
) -> String {
    let x = base.x.unwrap_or(0.0);
    let y = base.y.unwrap_or(0.0);
    let w = base.width.unwrap_or(100.0);

    let off_x: f64 = offset_x.trim_end_matches("mm").trim().parse().unwrap_or(0.0);
    let off_y: f64 = offset_y.trim_end_matches("mm").trim().parse().unwrap_or(0.0);

    let mut out = String::new();
    if base.page_break_before.unwrap_or(false) {
        out.push_str("#pagebreak(weak: true)\n");
    }
    out.push_str(&format!(
        "{}align(top + left)[#pad(top: {}mm, left: {}mm)[#block(width: {}mm, clip: false)[{}]]]\n",
        prefix,
        off_y + y,
        off_x + x,
        w,
        body
    ));
    out
}
